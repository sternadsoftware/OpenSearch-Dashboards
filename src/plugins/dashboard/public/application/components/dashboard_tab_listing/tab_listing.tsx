/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiTab, EuiTabs } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { useOpenSearchDashboards } from '../../../../../opensearch_dashboards_react/public';
import { DashboardServices } from '../../../types';

interface DashboardTabListingProps {
  dashboards: any[];
  selectedDashboardId: string;
}

export interface TabListingConfig {
  groups: Group[];
}

interface Group {
  dashboardId: string;
  detailDashboards: string[];
}

const config: TabListingConfig = {
  groups: [
    {
      dashboardId: 'c39012d0-eb7a-11ed-8e00-17d7d50cd7b2',
      detailDashboards: [
        'edf84fe0-e1a0-11e7-b6d5-4dc382ef7f5b',
        '722b74f0-b882-11e8-a6d9-e546fe2bba5f',
      ],
    },
    {
      dashboardId: '7adfa750-4c81-11e8-b3d7-01146121b73d',
      detailDashboards: ['722b74f0-b882-11e8-a6d9-e546fe2bba5f'],
    },
  ],
};

export const DashboardTabListing = ({
  dashboards,
  selectedDashboardId,
}: DashboardTabListingProps) => {
  const [selectedGroupId, setSelectedGroupId] = useState<string | undefined>(undefined);
  const [selectedDetailsDashboard, setSelectedDetailsDashboard] = useState<string | undefined>(
    undefined
  );

  const { services } = useOpenSearchDashboards<DashboardServices>();

  const isDashboardSelected = (dashboardId: string): boolean => {
    return selectedGroupId !== undefined && selectedGroupId === dashboardId;
  };

  const isDetailsDashboardSelected = (dashboardId: string): boolean => {
    return selectedDetailsDashboard !== undefined && selectedDetailsDashboard === dashboardId;
  };

  const selectDashboard = (dashboardId: string, isDetailsBoard: boolean = false) => {
    if (!isDetailsBoard) {
      setSelectedGroupId(dashboardId);
      setSelectedDetailsDashboard(undefined);
    } else {
      setSelectedDetailsDashboard(dashboardId);
    }
    services.application.navigateToUrl(`#/view/${dashboardId}`);
    const container = document.getElementById('dashboardViewport')?.firstChild;
    if (container) {
      // todo tlongo is this the right way to do this?
      ReactDOM.unmountComponentAtNode(document.getElementById('dashboardViewport') as Element);
    }
  };

  useEffect(() => {
    setSelectedGroupId(selectedDashboardId);
  }, [selectedDashboardId]);

  const renderDashboardTab = (dashboard: any, isDetailsBoard: boolean = false) => {
    return (
      <EuiTab
        isSelected={
          isDetailsBoard
            ? isDetailsDashboardSelected(dashboard!.id)
            : isDashboardSelected(dashboard!.id)
        }
        key={dashboard!.id}
        onClick={() => selectDashboard(dashboard!.id, isDetailsBoard)}
      >
        {dashboard!.title}
      </EuiTab>
    );
  };

  return (
    <>
      <EuiTabs>
        {dashboards &&
          config.groups.map((group) => {
            const dashboard = dashboards.find((d) => d.id === group.dashboardId);

            return renderDashboardTab(dashboard);
          })}
      </EuiTabs>
      <EuiTabs display="condensed" size="s">
        {dashboards &&
          selectedGroupId &&
          config.groups
            .filter((group) => group.dashboardId === selectedGroupId)
            .flatMap((group) => {
              return group.detailDashboards.map((dashboardId) => {
                const dashboard = dashboards.find((d) => d.id === dashboardId);

                return renderDashboardTab(dashboard, true);
              });
            })}
      </EuiTabs>
    </>
  );
};
