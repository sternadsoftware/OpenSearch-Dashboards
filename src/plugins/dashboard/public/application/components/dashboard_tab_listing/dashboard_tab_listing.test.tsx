/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { DashboardTabListing } from './tab_listing';
import { mount } from 'enzyme';

const dashboards = [
  {
    id: 'c39012d0-eb7a-11ed-8e00-17d7d50cd7b2',
    title: 'One',
  },
  {
    id: 'edf84fe0-e1a0-11e7-b6d5-4dc382ef7f5b',
    title: 'Two',
  },
  {
    id: '722b74f0-b882-11e8-a6d9-e546fe2bba5f',
    title: 'Three',
  },
  {
    id: '7adfa750-4c81-11e8-b3d7-01146121b73d',
    title: 'Three',
  },
];

const renderListing = () => {
  return (
    <DashboardTabListing
      dashboards={dashboards}
      selectedDashboardId="c39012d0-eb7a-11ed-8e00-17d7d50cd7b2"
    />
  );
};

describe('DashboardTabListing', () => {
  it('should render', () => {
    const component = mount(renderListing());
    component.update();

    expect(component).toMatchSnapshot();
  });
});
