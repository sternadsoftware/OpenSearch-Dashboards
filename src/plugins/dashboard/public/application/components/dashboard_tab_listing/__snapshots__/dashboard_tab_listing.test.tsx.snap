// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardTabListing should render 1`] = `
<DashboardTabListing
  dashboards={
    Array [
      Object {
        "id": "c39012d0-eb7a-11ed-8e00-17d7d50cd7b2",
        "title": "One",
      },
      Object {
        "id": "edf84fe0-e1a0-11e7-b6d5-4dc382ef7f5b",
        "title": "Two",
      },
      Object {
        "id": "722b74f0-b882-11e8-a6d9-e546fe2bba5f",
        "title": "Three",
      },
      Object {
        "id": "7adfa750-4c81-11e8-b3d7-01146121b73d",
        "title": "Three",
      },
    ]
  }
  selectedDashboardId="c39012d0-eb7a-11ed-8e00-17d7d50cd7b2"
>
  <EuiTabs>
    <div
      className="euiTabs"
      role="tablist"
    >
      <EuiTab
        isSelected={true}
        key="c39012d0-eb7a-11ed-8e00-17d7d50cd7b2"
        onClick={[Function]}
      >
        <button
          aria-selected={true}
          className="euiTab euiTab-isSelected"
          disabled={false}
          onClick={[Function]}
          role="tab"
          type="button"
        >
          <span
            className="euiTab__content"
          >
            One
          </span>
        </button>
      </EuiTab>
      <EuiTab
        isSelected={false}
        key="7adfa750-4c81-11e8-b3d7-01146121b73d"
        onClick={[Function]}
      >
        <button
          aria-selected={false}
          className="euiTab"
          disabled={false}
          onClick={[Function]}
          role="tab"
          type="button"
        >
          <span
            className="euiTab__content"
          >
            Three
          </span>
        </button>
      </EuiTab>
    </div>
  </EuiTabs>
  <EuiTabs
    display="condensed"
    size="s"
  >
    <div
      className="euiTabs euiTabs--condensed euiTabs--small"
      role="tablist"
    >
      <EuiTab
        isSelected={false}
        key="edf84fe0-e1a0-11e7-b6d5-4dc382ef7f5b"
        onClick={[Function]}
      >
        <button
          aria-selected={false}
          className="euiTab"
          disabled={false}
          onClick={[Function]}
          role="tab"
          type="button"
        >
          <span
            className="euiTab__content"
          >
            Two
          </span>
        </button>
      </EuiTab>
      <EuiTab
        isSelected={false}
        key="722b74f0-b882-11e8-a6d9-e546fe2bba5f"
        onClick={[Function]}
      >
        <button
          aria-selected={false}
          className="euiTab"
          disabled={false}
          onClick={[Function]}
          role="tab"
          type="button"
        >
          <span
            className="euiTab__content"
          >
            Three
          </span>
        </button>
      </EuiTab>
    </div>
  </EuiTabs>
</DashboardTabListing>
`;
